model/job-management/src/main/proto/beam_job_api.proto

```diff
@@ -201,6 +201,16 @@ message JobMessagesResponse {
 }
 
 // Enumeration of all JobStates
+//
+// The state transition diagram is:
+//   STOPPED -> STARTING -> RUNNING -> DONE
+//                                  \> FAILED
+//                                  \> CANCELLING -> CANCELLED
+//                                  \> UPDATING -> UPDATED
+//                                  \> DRAINING -> DRAINED
```

Comment:
- can we mark here which states are terminal with a new node (e.g. `-> [terminal]`)?   
- IIUC, `UPDATED` is not terminal.  What are the valid state transitions from `UPDATED`?
- the description for the `STOPPED` enum is "The job has been paused, or has not yet started."  That implies that a job can transition from `RUNNING -> STOPPED`.  what are the valid transitions from there?  My guess would be `STARTING`, `DONE`, and `CANCELING`, but for simplicity maybe it's just `STARTING`.

Reply:
Here's my proposal:

```
// The state transition diagram is:
//   STOPPED -> STARTING -> RUNNING -> DONE
//                                  \> FAILED
//                                  \> CANCELLING -> CANCELLED
//                                  \> UPDATING -> UPDATED -> RUNNING
//                                  \> DRAINING -> DRAINED
//                                  \> STOPPED -> STARTING
```

I decided to ditch the terminal idea.  It's implied by the lack of any further transitions.

Reply:
Yeah, this is the problem with UPDATED since it is dependent on the implementation within the Runner. Is an UPDATED job the same job as the prior one or a new job which continues from the prior state? In Dataflow it is the latter but the former can make sense for other runners.

Not sure of any RUNNING -> STOPPED scenarios today but adding

Reply:
> Not sure of any RUNNING -> STOPPED scenarios today.

This is either a pause-like event or an  erroneous use of STOPPED:

https://github.com/apache/beam/blob/master/runners/spark/src/main/java/org/apache/beam/runners/spark/SparkPipelineResult.java#L133

Reply:
That should either be DONE, FAILED, or CANCELLED

Reply:
Is there someone we can loop into this conversation who would know?

Reply:
Flink does not support updating existing jobs. It will always be a new job which continues from a checkpoint. Same for Spark. So `RUNNING -> STOPPED -> STARTED` seems hypothetical, also because there is no API for it currently, e.g. in `PipelineResult`.

Reply:
Ok, it sounds like we can pass on `RUNNING -> STOPPED -> STARTED` and `UPDATED -> RUNNING` for now.

I created a new Jira issue for the Spark problem: https://issues.apache.org/jira/browse/BEAM-8552

I'll handle the python runner issues in my PR.

The last thing I think we need to answer is whether UPDATED should be added to the list of default terminal states in Java.  It sounds like it is terminal in most cases right now, so I'd say yes.

Reply:
Should add here that it is possible for jobs to transition from `FAILING` to `RUNNING` in Flink (i.e. recovering from a checkpoint), though this does not have to be exposed through the Beam API.